app/adapters/osz_archive.py

```
@@ -0,0 +1,242 @@
+    carry a new name and new text). Every other member is copied through as raw
+    bytes in its original order, so audio/background/storyboard references inside
+    the difficulties keep resolving.
+
+    The replacement bytes are exactly ``text.encode("utf-8")`` -- the same bytes
+    the caller writes to ``.data/osu/{id}.osu`` -- which is what lets the md5 in
+    the database describe both copies.
+    """
+    buffer = io.BytesIO()
+
+    with _open_archive(data) as source:
+        with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as target:
```

Ruff (0.16.2)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)